ResponseTimes/Program.fs

```diff
@@ -1,51 +1,47 @@
-﻿// Learn more about F# at http://fsharp.org
-// See the 'F# Tutorial' project for more help.
-open LogParser
+﻿open LogParser
 open System
 open System.IO
+open Nessos.Streams
 
 let removeEverythingAfterLast (separator : char) (input : string) =
-    let parts = input.Split([|separator|], StringSplitOptions.RemoveEmptyEntries)
-    let numberOfItems = parts.Length 
+    let parts = input.Split([| separator |], StringSplitOptions.RemoveEmptyEntries)
+    let numberOfItems = parts.Length
     let strSeparator = separator.ToString()
 
-    Seq.take (numberOfItems - 1) parts
-    |> Seq.toList
+    parts
+    |> Array.take (numberOfItems - 1)
     |> String.concat strSeparator
-    |> (fun result -> strSeparator + result)
+    |> fun result -> strSeparator + result
 
 let cleanseRequestedUrl (url : string) =
-    let lower = url.ToLower()
-                    .Replace("/sams", "")
-                    .Replace("/service", "")
-                    .Replace("/ess", "")
-
-    if (lower.Contains "api/v1/documents/" || lower.Contains "api/v1/eftaccounts/")  then
-        removeEverythingAfterLast '/' lower
-    else lower
-
-let averageResponseTimesByRequestedURL logDetails =
-    Seq.filter(fun logDetail -> logDetail.ResponseTime.IsSome && logDetail.ResponseTime.Value > 0) logDetails
-    |> Seq.groupBy(fun logDetail -> sprintf "%s,%s" logDetail.HttpVerb (cleanseRequestedUrl logDetail.RequestedURL.AbsolutePath)) 
-    |> Seq.map(fun (key, group) -> key, group |> Seq.map(fun logDetail -> float logDetail.ResponseTime.Value) |> Seq.average)
-    |> Seq.sortByDescending(fun (key, value) -> value)
-
-
-
+    let url = url.ToLower()
+                 .Replace("/sams", "")
+                 .Replace("/service", "")
+                 .Replace("/ess", "")
+
+    if (url.Contains "api/v1/documents/" || url.Contains "api/v1/eftaccounts/") then removeEverythingAfterLast '/' url
+    else url
+
+let averageResponseTimesByRequestedURL =
+    ParStream.choose(fun (logDetail:LogDetailsStruct) ->
+        match logDetail.ResponseTime with
+        | Some responseTime when responseTime > 0 -> Some(responseTime, logDetail.HttpVerb, logDetail.RequestedURL)
+        | _ -> None)
+    >> ParStream.groupBy(fun (_, verb, url) -> verb, cleanseRequestedUrl url.AbsolutePath)
```

Comment:
Exactly. Tuples themselves support structural equality, as do verb and url so it should just work. Not only does this make "more sense" to me but should perform better than building a string.